spec/src/main/asciidoc/architecture.adoc

```diff
@@ -142,6 +142,8 @@ export MP_METRICS_TAGS=app=shop,tier=integration,special=deli\=ver\,y
 
 Global tags and tags registered with the metric are included in the output returned from the REST API.
 
+Global tags MUST NOT be added to the `MetricID` objects. Global tags must be included in list of tags when metrics are exported.
```

Comment:
In trying to implement this on our multi-app Open Liberty server, we're noticing
- we intend to add the _app tag at registration time
- MP Config currently allows config values to be application specific, which means the "global tags" are potentially app-specific tags.
- if we are going to add the app tag to the `MetricID` at the time of registration and want to continue to support app-specific global tags (because there's no real reason not to), it would make sense to put the "global" tag values into the MetricID as well.

So question is -- is there any reason not to allow all of the config-sourced tags (global and _app) to be put into the MetricID at registration time?  If not, we should remove the sentence above that says global tags must not be added to the MetricID objects.

Reply:
I think you just stop reading global tags within an app context.. ignore them. Global tags are global tags. If you want some other kind of app-common tags, then do that separately.

Reply:
Yeah, reading global tags in the context of an application and thus potentially allowing different apps to have different global tags, that sounds wrong.

Reply:
Fair enough... @Channyboy was looking last night to see if we could call MP Config to get the global tags purposely without any application context.  At least at first glance it appears that we can do that as long as we choose the right classloader (ie. not the TCCL) to pass to MP Config to do the lookup.  We may want to clarify in the spec that we explicitly don't want application context config for the global tags.

Reply:
@donbourne applied that via https://github.com/eclipse/microprofile-metrics/pull/595/commits/46f71199819a1f82f1488ef9a36b270d09530c84, looks good?

Reply:
Hmm... I think we may need (conceptual) separation between global tags (apply to all metrics across all registries within the process), and registry-level tags (apply to all metrics within a single registry).

The former can be easily mapped to latter (and in most applications they will likely be one and the same, if we switch to a single registry per #384 discussion), but there may be situations when global tags may map to multiple registries (such as multi-app servers @donbourne described), and there may be situations where you want per registry tags that are not truly global (such as app name discussed earlier).

I don't think this impacts this PR per se, but it is something that should be defined by the spec.

I do think this PR should add support for registry-level tag management in order to be complete, though.

Reply:
micrometer does this because of it's composed registry structure. Same mechanism is used (registry.config().commonTags() is one way, there are others), just depends on which registry you apply it to (global/top-level composite, or an individual registry). Significant difference is that a registry in micrometer is more closely aligned to an exporter in MP Metrics instead of being associated with a scope. I think someone needs to explain to me the value of scopes.. because outside of strictly hierarchical names, I don't get it (not a conversation for this item, however. ;) )

Reply:
I'd say that if an implementation needs registry-scoped tags, it's not a problem to add that as a vendor-specific thing, but I wouldn't bake it into the spec, especially when we're considering switching to a single registry

Reply:
> I'd say that if an implementation needs registry-scoped tags, it's not a problem to add that as a vendor-specific thing, but I wouldn't bake it into the spec, especially when we're considering switching to a single registry

I disagree -- that (an optional, "vendor-specific" way of defining tags at the registry level) is exactly what I want to avoid.

If that happens, the best case scenario is that I have to implement different integration for each MP implementation, which is both cumbersome and pointless. Worst case scenario is that I can't add the tags I need to add because the vendor didn't provide a way for me to do it to begin with.

Again, I am not an MP vendor, and while I may be able to influence one MP vendor (Helidon), I can't really influence what anyone else does. This feature has to be defined at the spec level so I, and any other spec **user** who may need it, can rely on it to be able to add our "global" tags programmatically to one or more registries.

Switching from multiple to a single registry doesn't really change that need, btw. It just makes it easier for me to implement if there is a single registry to deal with.

Reply:
Could you send a separate PR for that if you think it's needed? I don't have a clear idea about how exactly you want to do this. I'd say let's get this one merged and not over-complicate it.